src/rust/kafka/src/test_utils/topic_scanner.rs

```diff
@@ -9,133 +13,184 @@ use rust_proto::{
+    pub fn new(consumer_config: ConsumerConfig, timeout: Duration) -> Self {
+        // this overridden config adds a unique suffix to the consumer group,
+        // ensuring that each test consumer belongs to its own unique group
+        let overridden_config = ConsumerConfig {
+            bootstrap_servers: consumer_config.bootstrap_servers,
+            sasl_username: consumer_config.sasl_username,
+            sasl_password: consumer_config.sasl_password,
+            consumer_group_name: format!(
+                "{}-{}",
+                consumer_config.consumer_group_name,
+                uuid::Uuid::new_v4(),
+            ),
+            topic: consumer_config.topic,
+        };
+
+        Self {
+            consumer_config: overridden_config,
+            timeout,
+            t_: PhantomData,
+        }
     }
 
+    /// Consume messages from Kafka matching the given tenant_id, filtered by
+    /// the predicate. This will terminate after max_envelopes have been
+    /// consumed. Panics if this takes longer than self.timeout.
+    ///
+    /// N.B.: If fewer than max_envelopes messages matching the tenant_id and
+    /// predicate are available, this will time out.
     pub async fn contains_for_tenant(
         self,
         tenant_id: uuid::Uuid,
+        max_envelopes: usize,
         mut predicate: impl FnMut(T) -> bool + Send + Sync + 'static,
-    ) -> Result<
-        ScanReadyToGetResult<Result<Envelope<T>, KafkaTopicContainsError>>,
-        KafkaTopicContainsError,
-    > {
+    ) -> JoinHandle<Vec<Envelope<T>>> {
         let tenant_eq_predicate = move |envelope: Envelope<T>| {
             let envelope_tenant_id = envelope.tenant_id();
             let inner_message = envelope.inner_message();
             envelope_tenant_id == tenant_id && predicate(inner_message)
         };
-        self.contains(tenant_eq_predicate).await
+        self.contains(tenant_eq_predicate, move |idx, _envelope| {
+            idx >= max_envelopes
+        })
+        .await
     }
 
+    /// Consume messages from Kafka matching filter_predicate into a list until
+    /// the stop_predicate returns true. Panics if this takes longer than
+    /// self.timeout.
     pub async fn contains(
         self,
-        predicate: impl FnMut(Envelope<T>) -> bool + Send + Sync + 'static,
-    ) -> Result<
-        ScanReadyToGetResult<Result<Envelope<T>, KafkaTopicContainsError>>,
-        KafkaTopicContainsError,
-    > {
+        filter_predicate: impl FnMut(Envelope<T>) -> bool + Send + Sync + 'static,
+        stop_predicate: impl FnMut(usize, Envelope<T>) -> bool + Send + Sync + 'static,
+    ) -> JoinHandle<Vec<Envelope<T>>> {
         // we'll use this channel to communicate that the consumer is ready to
         // consume messages
         let (tx, rx) = oneshot::channel::<()>();
-
-        let predicate = std::sync::Arc::new(std::sync::Mutex::new(predicate));
+        let tx_mutex = Mutex::new(Some(tx));
```

Comment:
gonna skip reviewing this part re: rx-tx, but the stop/filter predicate _is_ cool

Reply:
i guess `contains` is no longer the right term though, maybe just `scan` or `filter`. we can fix that up later though

Reply:
fixed